marketdata/coingecko/client.go

```diff
@@ -0,0 +1,82 @@
+package coingecko
+
+import (
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+	"github.com/trustwallet/blockatlas/pkg/logger"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+)
+
+type Client struct {
+	blockatlas.Request
+}
+
+func NewClient(api string) *Client {
+	c := Client{
+		Request: blockatlas.InitClient(api),
+	}
+	return &c
+}
+
+func (c *Client) FetchLatestRates(coins GeckoCoins) (prices CoinPrices) {
+	ci := coinIds(coins)
+
+	bucketSize := 500
+	i := 0
+	prChan := make(chan CoinPrices)
+	var wg sync.WaitGroup
+	for i < len(ci) {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			var end = len(ci)
+			if len(ci) > i+bucketSize {
+				end = i + bucketSize
+			}
+			bucket := ci[i:end]
+			values := url.Values{
+				"vs_currency": {blockatlas.DefaultCurrency},
+				"sparkline":   {"false"},
+				"ids":         {strings.Join(bucket[:], ",")},
+			}
+
+			var cp CoinPrices
+			err := c.Get(&cp, "v3/coins/markets", values)
+			if err != nil {
+				logger.Error(err)
+				return
+			}
+			prChan <- cp
+		}(i)
+
+		i += bucketSize
+	}
+
+	go func() {
```

Comment:
see Fan-out, fan-in in the official Go docs

https://blog.golang.org/pipelines